fl4health/model_bases/feature_extractor_buffer.py

```diff
@@ -0,0 +1,158 @@
+                extraction layers.
+            fhooks (List[RemovableHandle]): A list to store the handles for removing hooks.
+            accumulate_features (bool): A flag indicating whether to accumulate features.
+            extracted_features_buffers (Dict[str, List[torch.Tensor]]): A dictionary to store the extracted features
+                for each layer.
+        """
+        self.model = model
+        self.flatten_feature_extraction_layers = flatten_feature_extraction_layers
+        self.fhooks: List[RemovableHandle] = []
+
+        self.accumulate_features: bool = False
+        self.extracted_features_buffers: Dict[str, List[torch.Tensor]] = {
+            layer: [] for layer in flatten_feature_extraction_layers.keys()
+        }
+
+    def enable_accumulating_features(self) -> None:
+        """
+        Enables the accumulation of features in the buffers for multiple forward passes.
+
+        This method sets the `accumulate_features` flag to True, allowing the model to accumulate features
+        in the buffers for multiple forward passes. This can be useful in scenarios where you want to extract
+        features from intermediate layers of the model during inference.
+        """
+        self.accumulate_features = True
+
+    def disable_accumulating_features(self) -> None:
+        """
+        Disables the accumulation of features in the buffers.
+
+        This method sets the `accumulate_features` attribute to False, which prevents the buffers from accumulating
+        features and overwrits them for each forward pass.
+        """
+        self.accumulate_features = False
+
+    def clear_buffers(self) -> None:
+        """
+        Clears the extracted features buffers for all layers.
+        """
+        self.extracted_features_buffers = {layer: [] for layer in self.flatten_feature_extraction_layers.keys()}
+
+    def get_hierarchical_attr(self, module: nn.Module, layer_hierarchicy: List[str]) -> nn.Module:
+        if len(layer_hierarchicy) == 1:
+            return getattr(module, layer_hierarchicy[0])
+        else:
+            return self.get_hierarchical_attr(getattr(module, layer_hierarchicy[0]), layer_hierarchicy[1:])
+
+    def _maybe_register_hooks(self) -> None:
+        """
+        Checks if hooks are already registered and registers them if not.
+        Hooks extract the intermediat feature as output of the selected layers in the model.
+        """
+        if len(self.fhooks) == 0:
+            log(INFO, "Starting to register hooks:")
+            named_layers = dict(self.model.named_modules())
+            for layer in named_layers.keys():
+                if layer in self.flatten_feature_extraction_layers.keys():
+                    # Split the layer name by '.' to get the hirarchial attribute
+                    log(INFO, f"Registering hook for layer: {layer}")
+                    layer_hierarchicy = layer.split(".")
```

Comment:
Because I want to users be able to give more top level modules rather than specific name of module, I need to traverse to find the actual module function to apply this hook. Hooks can't be applied on top level models.

Reply:
Ah, right that makes sense! As long as we implement the test for the hierarchical search, this seems good to me!